File: openaev-framework/src/main/java/io/openaev/model/inject/form/Expectation.java
package io.openaev.model.inject.form;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.openaev.database.model.BaseInjectExpectation;
import io.openaev.database.model.SecurityPlatform;
import java.util.ArrayList;
import java.util.List;
import lombok.Data;

/**
 * Form model representing expectation configuration from user input.
 *
 * <p>This class captures expectation parameters as provided by users through the injection form. It
 * is used as input to create the appropriate expectation instances (Detection, Prevention, Manual,
 * etc.).
 *
 * <p>This is a data transfer object (DTO) that is deserialized from JSON form submissions and then
 * used to construct domain expectation objects.
 *
 * @see io.openaev.model.Expectation
 * @see io.openaev.model.expectation.DetectionExpectation
 * @see io.openaev.model.expectation.PreventionExpectation
 * @see io.openaev.expectation.ExpectationBuilderService
 */
@Data
@JsonAutoDetect(
    getterVisibility = JsonAutoDetect.Visibility.NONE,
    isGetterVisibility = JsonAutoDetect.Visibility.NONE)
public class Expectation {

  /** The type of expectation to create. */
  @JsonProperty("expectation_type")
  private BaseInjectExpectation.EXPECTATION_TYPE type;

  /** Display name for this expectation. */
  @JsonProperty("expectation_name")
  private String name;

  /** Detailed description of the expectation. */
  @JsonProperty("expectation_description")
  private String description;

  /** The score value when this expectation is fulfilled (0-100). */
  @JsonProperty("expectation_score")
  private Double score;

  /** Whether this expectation should be evaluated as part of a group. */
  @JsonProperty("expectation_expectation_group")
  private boolean expectationGroup;

  /** Time in seconds after which this expectation automatically expires. */
  @JsonProperty("expectation_expiration_time")
  private Long expirationTime;

  /**
   * Whether this expectation type is limited to one selection per inject.
   *
   * <p>Only relevant for available expectations exposed in the injector contract. MANUAL
   * expectations are not limited (can be added multiple times); all other types are limited to a
   * single selection.
   */
  @JsonProperty("expectation_is_multi_selectable")
  private boolean multiSelectable;

  @JsonProperty("expectation_is_predefined")
  private boolean predefined;

  /**
   * Optional display order of this expectation within its inject, ascending. Lets a contract
   * declare the logical sequence of its expectations (e.g. a phishing action orders its human steps
   * email {@literal ->} link {@literal ->} submission) instead of relying on an incidental
   * alphabetical sort. {@code null} means unordered - the reader then falls back to name / id, so
   * every other contract is unaffected.
   */
  @JsonProperty("expectation_order")
  private Integer order;

  /**
   * Security platform types expected to fulfil this expectation.
   *
   * <p>When non-empty, the platform focuses the detection/prevention result on collectors of those
   * types only (instead of every connected security platform). An empty list means "any security
   * platform" (unchanged behaviour), and is typical for MANUAL expectations.
   */
  @JsonProperty("expectation_expected_security_platform_types")
  private List<SecurityPlatform.SECURITY_PLATFORM_TYPE> expectedSecurityPlatformTypes =
      new ArrayList<>();
}

File: openaev-front/src/admin/components/common/injects/expectations/Expectation.ts
import { SECURITY_PLATFORM_TYPE_LABELS, type SecurityPlatformType } from '../../../../../components/securityPlatformType';
import { type InjectExpectationOutput } from '../../../../../utils/api-types';

export interface InjectExpectationsStore extends Omit<InjectExpectationOutput, 'inject_expectation_team' | 'inject_expectation_user' | 'inject_expectation_article' | 'inject_expectation_challenge' | 'inject_expectation_asset'> {
  inject_expectation_id: string;
  inject_expectation_type: InjectExpectationOutput['inject_expectation_type'];
  inject_expectation_expected_score: number;
  inject_expectation_expiration_time: number;
  inject_expectation_team: string | undefined;
  inject_expectation_user: string | undefined;
  inject_expectation_article: string | undefined;
  inject_expectation_challenge: string | undefined;
  inject_expectation_asset: string | undefined;
}

export interface ExpectationInput {
  expectation_type: string;
  expectation_name: string;
  expectation_description?: string;
  expectation_score: number;
  expectation_expectation_group: boolean;
  expectation_expiration_time: number;
  expectation_is_multi_selectable?: boolean;
  // Security platform types expected to fulfil this expectation (empty = any platform).
  expectation_expected_security_platform_types?: string[];
  expectation_is_predefined: boolean;
  // Display order within the inject, declared by the injector contract (e.g. phishing orders its
  // steps email -> link -> submission). Not user-editable: carried through edits so the results
  // timeline keeps its logical order. Null/undefined means unordered.
  expectation_order?: number | null;
}

// Derived from the canonical type map so a platform type added server-side
// (e.g. VULNERABILITY_SCANNER) always shows up in the expectation forms.
export const SECURITY_PLATFORM_TYPES = Object.keys(SECURITY_PLATFORM_TYPE_LABELS) as SecurityPlatformType[];

// expectation_order is deliberately not part of the form: it is contract-declared, not
// user-editable, and ExpectationPopover carries it through on submit.
export interface ExpectationInputForm extends Omit<ExpectationInput, 'expectation_expiration_time' | 'expectation_is_multi_selectable' | 'expectation_order'> {
  expiration_time_days: number;
  expiration_time_hours: number;
  expiration_time_minutes: number;
}

export enum ExpectationType {
  PREVENTION = 'PREVENTION',
  DETECTION = 'DETECTION',
  VULNERABILITY = 'VULNERABILITY',
  MANUAL = 'MANUAL',
  ARTICLE = 'ARTICLE',
  CHALLENGE = 'CHALLENGE',
}

export type ExpectationResultType = 'PREVENTION' | 'DETECTION' | 'VULNERABILITY' | 'HUMAN_RESPONSE';

export const expectationResultTypes: ExpectationResultType [] = ['PREVENTION', 'DETECTION', 'VULNERABILITY', 'HUMAN_RESPONSE'];

export const mitreMatrixExpectationTypes: ExpectationResultType [] = ['PREVENTION', 'DETECTION', 'VULNERABILITY'];
